provider/provider_yaml_test.go

```diff
@@ -452,14 +470,439 @@ func TestIMDSAuth(t *testing.T) {
+//  4. `Import` using the CLI. Ensures resources can be successfully imported
+//     4a. Allows for a hook to be run prior to import being run. e.g. Add tags remotely
+//  5. `Refresh` with no changes
+func testTagsPulumiLifecycle(t *testing.T, step tagsTestStep) {
+	t.Helper()
+	ctx := context.Background()
+
+	stepDir, err := os.MkdirTemp(os.TempDir(), step.name)
+	assert.NoError(t, err)
+	fpath := filepath.Join(stepDir, "Pulumi.yaml")
+
+	generateTagsTest(t, step, fpath, "")
+	ptest := pulumiTest(t, stepDir, opttest.TestInPlace())
+	stack := ptest.CurrentStack()
+
+	t.Log("Initial deployment...")
+	upRes, err := stack.Up(ctx)
+	assert.NoError(t, err)
+	outputs := upRes.Outputs
+	urn := outputs["urn"].Value.(string)
+	id := outputs["id"].Value.(string)
+	providerUrn := outputs["providerUrn"].Value.(string)
+	if step.postUpHook != nil {
+		step.postUpHook(t, outputs)
+	}
+
+	t.Log("refresh...")
+	_, err = stack.Refresh(ctx, optrefresh.ExpectNoChanges())
+	assert.NoError(t, err)
+
+	t.Log("delete state...")
+	execPulumi(t, ptest, stepDir, "state", "delete", urn)
+
+	// import using the import resource option
+	t.Log("up with import...")
+	if step.preImportHook != nil {
+		step.preImportHook(t, outputs)
+	}
+	generateTagsTest(t, step, fpath, id)
+	upRes, err = stack.Up(ctx, optup.Diff())
+	assert.NoError(t, err)
+	changes := *upRes.Summary.ResourceChanges
+	assert.Equal(t, 1, changes["import"])
+
+	t.Log("delete state...")
+	execPulumi(t, ptest, stepDir, "state", "delete", urn)
+
+	t.Log("import from cli...")
+	if step.preImportHook != nil {
+		step.preImportHook(t, outputs)
+	}
+	generateTagsTest(t, step, fpath, "")
+	execPulumi(t, ptest, stepDir, "import", step.typ, "res", id, "--provider", fmt.Sprintf("aws-provider=%s", providerUrn), "--yes")
+	execPulumi(t, ptest, stepDir, "state", "unprotect", urn, "--yes")
+
+	// need to run an up to fix the state. It should be a no-op
+	// re https://github.com/pulumi/pulumi-aws/issues/4204
+	upRes, err = stack.Up(ctx)
+	assert.NoError(t, err)
+	for k := range *upRes.Summary.ResourceChanges {
+		if k != "same" {
+			t.Fatal("expected no changes")
+		}
+	}
+
+	t.Log("preview with refresh...")
+	_, err = stack.Preview(ctx, optpreview.Refresh(), optpreview.ExpectNoChanges())
+	assert.NoError(t, err)
+}
+
+// generateTagsTest generates a pulumi program for the given test step
+// and writes it to the test directory
+func generateTagsTest(t *testing.T, step tagsTestStep, testPath string, importId string) {
+	template := `name: test-aws-%s
+runtime: yaml
+resources:
+  aws-provider:
+    type: pulumi:providers:aws%s%s
+  res:
+    type: %s%s%s
+outputs:
+  actual: ${res.tags}
+  urn: ${res.urn}
+  id: ${res.id}
+  resArn: ${res.arn}
+  providerUrn: ${aws-provider.urn}`
+
+	options := map[string]interface{}{
+		"provider": "${aws-provider}",
+	}
+
+	if importId != "" {
+		options["import"] = importId
+	}
+
+	var expandMap func(level int, v interface{}) string
+	expandMap = func(level int, v interface{}) string {
```

Comment:
Oof too bad there is no existing helper for this. Theoretically YAML is a text representation of json-like trees and could operate at that level perhaps followed by yaml.Marshal.